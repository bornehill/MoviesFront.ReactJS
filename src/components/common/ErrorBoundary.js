import React from "react";

class ErrorBoundary extends React.Component {
	constructor(props) {
		super(props);
		this.state = { error: null, errorInfo: null };
	}

	componentDidCatch(error, errorInfo) {
		this.setState({
			error: error,
			errorInfo: errorInfo,
		});
	}

	render() {
		if (this.state.errorInfo) {
			return (
				<div>
					<h2 className="font-emphasis text-flame-600">
						Something went wrong.
					</h2>
					<details className="whitespace-pre-wrap">
						{this.state.error && this.state.error.toString()}
						<br />
						{this.state.errorInfo.componentStack}
					</details>
				</div>
			);
		}

		return this.props.children;
	}
}

export default ErrorBoundary;
