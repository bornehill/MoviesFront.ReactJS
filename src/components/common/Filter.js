import React from "react";

const Filter = ({ sortBy, onSort, onFilter }) => {
	return (
		<div className="flex justify-between bg-movie-gray mt-5">
			<nav className="cursor-pointer px-5 pt-2">
				<ul className="flex text-white gap-5">
					<li onClick={() => onFilter("")}>ALL</li>
					<li onClick={() => onFilter("documentary")}>DOCUMENTARY</li>
					<li onClick={() => onFilter("comedy")}>COMEDY</li>
					<li onClick={() => onFilter("horror")}>HORROR</li>
					<li onClick={() => onFilter("crime")}>CRIME</li>
				</ul>
			</nav>
			<div className="pr-5 pt-2">
				<span className="text-movie-onyx mr-3">SORT BY</span>
				<select
					className="bg-movie-gray text-white"
					value={sortBy}
					onChange={(e) => onSort(e.target.value)}
				>
					<option value="release_date">RELEASE DATE</option>
					<option value="genres">GENRE</option>
					<option value="vote_average">RATING</option>
				</select>
			</div>
		</div>
	);
};

export default Filter;
