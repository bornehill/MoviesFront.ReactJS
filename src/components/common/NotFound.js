import React from "react";

const NotFound = () => {
	return (
		<>
			<h3 className="text-5xl text-movie-red text-center">Page not found</h3>
		</>
	);
};

export default NotFound;
