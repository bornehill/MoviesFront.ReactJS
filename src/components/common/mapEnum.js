export function mapEnumToDropdown(enumType) {
	return Object.keys(enumType).map((typeValue) => {
		return {
			value: typeValue,
			text: enumType[typeValue],
		};
	});
}

export function mapEnumToObject(enumType) {
	const enumObject = {};

	Object.keys(enumType).forEach((typeValue) => {
		enumObject[typeValue] = false;
	});

	return enumObject;
}
