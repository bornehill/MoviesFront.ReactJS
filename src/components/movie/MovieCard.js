import React, { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";

import MovieMenu from "./MovieMenu";
import { MovieContext } from "../../providers/Movie.provider";

const MovieCard = ({ movie, fnDelete, fnEdit }) => {
	const navigate = useNavigate();
	const [showMenu, setShowMenu] = useState(false);
	const { setMovieDetailId } = useContext(MovieContext);

	const displayMenu = (show) => {
		setShowMenu(show);
	};

	const showDetail = () => {
		window.scrollTo(0, 0);
		setMovieDetailId(movie.id);
		navigate(`/search?movie=${movie.id}`);
	};

	return (
		<div
			className="card card-white max-w-sm m-5"
			onMouseOver={() => displayMenu(true)}
			onMouseLeave={() => displayMenu(false)}
		>
			<MovieMenu
				show={showMenu}
				onDelete={() => fnDelete(movie.id)}
				onEdit={() => fnEdit(movie)}
			/>
			<img
				className="w-full rounded-t-sm"
				alt=""
				src={movie.poster_path}
				onClick={showDetail}
			/>
			<div className="px-4 py-2 flex justify-between">
				<div>
					<p className="font-light">{movie.title}</p>
					<p className="font-light">
						{movie.genres?.map((g) => {
							return g + ", ";
						})}
					</p>
				</div>
				<p className="border-2 border-movie-onyx p-1 font-light">
					{new Date(movie.release_date).getFullYear()}
				</p>
			</div>
		</div>
	);
};

export default MovieCard;
