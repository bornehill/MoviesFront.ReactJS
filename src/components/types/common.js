export const movieEmpty = {
	title: "",
	poster_path: "",
	genres: [],
	release_date: "",
	runtime: 0,
	overview: "",
	vote_average: 0,
};

export const Genres = {
	Drama: "Drama",
	Romance: "Romance",
	Fantasy: "Fantasy",
	Adventure: "Adventure",
	"Science Fiction": "Science Fiction",
	Comedy: "Comedy",
	Family: "Family",
	Animation: "Animation",
	Action: "Action",
};
