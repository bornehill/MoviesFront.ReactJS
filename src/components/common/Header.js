import React, { useState } from "react";

import { useAddMovieMutation } from "../../reducers/movieSlice";

import SuccessMsg from "../movie/SuccessMsg";
import FormVideo from "../movie/FormVideo";
import { movieEmpty } from "../../components/types/common";

const Header = () => {
	const [addMovie, { error }] = useAddMovieMutation();
	const [showMsg, setShowMsg] = useState(false);
	const [newMovie, setNewMovie] = useState();

	const closeMsg = () => {
		setShowMsg(false);
	};

	const CancelAdd = () => {
		setNewMovie();
	};

	const onAddMovie = () => {
		setNewMovie({ ...movieEmpty });
	};

	const ConfirmAdd = async (movie) => {
		movie.vote_average = +movie.vote_average;
		movie.runtime = +movie.runtime;

		await addMovie(movie);
		setNewMovie();
		if (!error) setShowMsg(true);
	};

	return (
		<>
			{showMsg && (
				<SuccessMsg
					title={"CONGRATULATIONS!"}
					subTitle={"The movie has been added to database successfully"}
					onCancel={closeMsg}
				/>
			)}
			{newMovie && (
				<FormVideo movie={newMovie} onCancel={CancelAdd} onEdit={ConfirmAdd} />
			)}
			{error && (
				<p className="mt-2 text-sm p-2 text-white bg-red-700">{error}</p>
			)}
			<header className="bg-movie-gray py-4">
				<div className="px-8 md:flex md:items-center md:justify-between">
					<div className="flex items-center">
						<span className="text-xl text-movie-red font-bold">netflix</span>
						<span className="font-light text-2xl text-movie-red">roulette</span>
					</div>
					<button
						className="btn button p-2 bg-movie-onyx text-movie-red font-light rounded"
						onClick={onAddMovie}
					>
						+ ADD MOVIE
					</button>
				</div>
				<h2 className="text-4xl font-light text-white mr-2 md:mx-32 py-4 md:py-3 px-2 md:px-0">
					FIND YOUR MOVIE
				</h2>
				<div className="max-w-full md:mx-32">
					<input
						id="search"
						className="text-white bg-movie-onyx mx-2 md:mr-4 md:w-2/3"
						placeholder="What do you what to watch?"
					/>
					<button
						id="btnSearch"
						className="text-white bg-movie-red md:w-1/4 p-1 md:py-2 rounded"
					>
						SEARCH
					</button>
				</div>
			</header>
		</>
	);
};

export default Header;
