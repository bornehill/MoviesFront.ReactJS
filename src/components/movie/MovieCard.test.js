import { render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { MovieProvider } from "../../providers/Movie.provider";
import MovieCard from "./MovieCard";
import { movieEmpty } from "../../components/types/common";
import { createMemoryHistory } from "history";

const handleDelete = jest.fn();
const handleEdit = jest.fn();
const testMovie = {
	...movieEmpty,
	id: 334536,
	title: "The Blackcoat's Daughter",
	release_date: "2015-11-20",
	runtime: 93,
	poster_path:
		"https://image.tmdb.org/t/p/w500/gW5MnkQ9zHtyVPzdDDHPEDCrZ57.jpg",
};

test("MovieCard should be showed and match snapshot", () => {
	const card = render(
		<BrowserRouter>
			<MovieProvider>
				<Routes>
					<Route
						path="*"
						element={
							<MovieCard
								movie={testMovie}
								fnDelete={handleDelete}
								fnEdit={handleEdit}
							/>
						}
					/>
				</Routes>
			</MovieProvider>
		</BrowserRouter>
	);

	expect(screen.getByText("The Blackcoat's Daughter")).toBeInTheDocument();
	expect(card).toMatchSnapshot();
});

test("MovieCard should show details", async () => {
	const history = createMemoryHistory();

	render(
		<BrowserRouter history={history}>
			<MovieProvider>
				<Routes>
					<Route
						path="*"
						element={
							<MovieCard
								movie={testMovie}
								fnDelete={handleDelete}
								fnEdit={handleEdit}
							/>
						}
					/>
				</Routes>
			</MovieProvider>
		</BrowserRouter>
	);

	const movieImg = screen.getByRole("img");
	userEvent.click(movieImg);
	await waitFor(() => expect(history.location.pathname).toBe("/"));
});
