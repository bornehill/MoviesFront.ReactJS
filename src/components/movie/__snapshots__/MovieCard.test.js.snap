// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should be showed and match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card card-white max-w-sm m-5"
      >
        <img
          alt=""
          class="w-full rounded-t-sm"
          src="https://image.tmdb.org/t/p/w500/gW5MnkQ9zHtyVPzdDDHPEDCrZ57.jpg"
        />
        <div
          class="px-4 py-2 flex justify-between"
        >
          <div>
            <p
              class="font-light"
            >
              The Blackcoat's Daughter
            </p>
            <p
              class="font-light"
            />
          </div>
          <p
            class="border-2 border-movie-onyx p-1 font-light"
          >
            2015
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card card-white max-w-sm m-5"
    >
      <img
        alt=""
        class="w-full rounded-t-sm"
        src="https://image.tmdb.org/t/p/w500/gW5MnkQ9zHtyVPzdDDHPEDCrZ57.jpg"
      />
      <div
        class="px-4 py-2 flex justify-between"
      >
        <div>
          <p
            class="font-light"
          >
            The Blackcoat's Daughter
          </p>
          <p
            class="font-light"
          />
        </div>
        <p
          class="border-2 border-movie-onyx p-1 font-light"
        >
          2015
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
